### PR TITLE
Added ConcurrentDictionary.Comparer property

### LINKED ISSUE #31350: Add ConcurrentDictionary.Comparer property
It would be useful to expose this as a getter property. I believe all other BCL collections that accept comparers expose them as properties (i.e. `Dictionary<,>.Comparer`, `HashSet<>.Comparer`).

### PR DESCRIPTION
**Pull request by I-SER-I**
fix #31350

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=47787) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=47787) before we can accept your contribution.<br/><br/>:x: I-SER-I [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=47787)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=47787) it.</sub>

**Comment by I-SER-I** [comment]
In [video ](https://youtu.be/mecKToY4JaE?t=319 ) said that it is a copy of the Dictionary. It no works cause interface IInternalStringEqualityComparer placed in CoreLIb. What to do?

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs [hunk 1]
```diff
@@ -1051,6 +1051,21 @@ private bool TryAddInternal(TKey key, int? nullableHashcode, TValue value, bool
...
+            {
+                if (typeof(TKey) == typeof(string))
+                {
+                    return (IEqualityComparer<TKey>)IInternalStringEqualityComparer.GetUnderlyingEqualityComparer((IEqualityComparer<string?>?)_comparer);
+                }
+                else
+                {
+                    return _comparer ?? EqualityComparer<TKey>.Default;
+                }
+            }
```
Comment (stephentoub) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:1066]: The entire implementation should just be:
```C#
public IEqualityComparer<TKey> Comparer => _comparer ?? _defaultComparer;
```

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs [hunk 2]
```diff
@@ -1051,6 +1051,21 @@ private bool TryAddInternal(TKey key, int? nullableHashcode, TValue value, bool
...
+        }
+
```
Comment (stephentoub) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:1068]: Please add tests. Thanks.

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: Shouldn't this be consistent with [what the regular dictionary is doing](https://github.com/dotnet/runtime/blob/2e0ae35947d29c8be6357f0cb95205921fbfa28d/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L172-L179)?

Comment (stephentoub) [reply src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs]: No.  Dictionary upgrades its default string comparer from a (corelib internal) non-randomized comparer to a randomized comparer, hence it needing special logic in Comparer to compensate and hide the implementation detail.  ConcurrentDictionary has no such implementation.

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs [hunk 3]
```diff
@@ -1051,6 +1051,8 @@ private bool TryAddInternal(TKey key, int? nullableHashcode, TValue value, bool
         private static void ThrowKeyNotFoundException(TKey key) =>
             throw new KeyNotFoundException(SR.Format(SR.Arg_KeyNotFoundWithKey, key.ToString()));
 
+        public IEqualityComparer<TKey> Comparer => _comparer ?? _defaultComparer;
```
Comment (stephentoub) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:1054]: Nit: please add an xml comment to help seed the documentation for the new member.

**Review by stephentoub — APPROVED** [review]
Thanks!

**Comment by stephentoub** [comment]
@eiriktsarpalis, the member still needed XML docs added.  Do you want to add that?

**Comment by eiriktsarpalis** [comment]
My bad, created a new PR adding docs.

**Pull request merged**